src/category_theory/lifting_properties.lean

```diff
@@ -0,0 +1,208 @@
+/-
+Copyright (c) 2021 Jakob Scholbach. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob Scholbach
+-/
+import category_theory.category
+import category_theory.arrow
+import category_theory.functor
+import category_theory.limits.preserves.shapes.terminal
+
+/-!
+# Lifting properties
+
+This file defines the lifting property of two arrows in a category and show basic properties of 
+this notion. 
+We also construct the subcategory consisting of those morphisms which have the right lifting 
+property with respect to arrows in a given diagram.
+
+## Main results
+- `has_lifting_property`: the definition of the lifting property
+- `iso_has_right_lifting_property`: any isomorphism satisfies the right lifting property (rlp)
+- `id_has_right_lifting_property`: any identity has the rlp
+- `right_lifting_property_initial_iff`: spells out the rlp against a map whose source is an 
+initial object
+- `right_lifting_subcat`: given a functor F : D ⥤ arrow C, we construct the subcategory of those 
+morphisms p in C that satisfy the rlp w.r.t. F(i), for any object i in D.
+
+## Tags
+lifting property
+-/
+
+namespace category_theory
+
+universes v u
+variables {C : Type u} [category.{v} C]
+variables {D : Type u} [category.{v} D]
+
+variables {X Y Z : C}
+
+/- The left lifting property of a morphism i vs. a morphism p. -/
+def has_lifting_property (i p : arrow C) : Prop := ∀ (sq : i ⟶ p), arrow.has_lift sq
+
+/- Any isomorphism has the right lifting property with respect to any map. 
+A    → X 
+↓i    ↓p≅
+B    → Y
+-/
+lemma iso_has_right_lifting_property (i : arrow C) (p : X ≅ Y) : 
+has_lifting_property i (arrow.mk p.hom) :=
+begin
+  intro sq,
+  fconstructor,
+  fconstructor,
+  fconstructor,
+  { use sq.right ≫ p.inv }, -- the lift is obtained by p⁻¹ ∘ (B → Y)
+  { rw [←category.assoc, (arrow.w sq).symm],
+    simp, },
+  { simp, }
+end
```

Comment:
This is perhaps a matter of taste, but you can also write this as 
```
-- the lift is obtained by p⁻¹ ∘ (B → Y)
has_lifting_property i (arrow.mk p.hom) := λ sq,
⟨⟨⟨sq.right ≫ p.inv, by { rw [←category.assoc, (arrow.w sq).symm], simp, }, by simp⟩⟩⟩
```

Reply:
Further, my inclination facing this proof would be to aspire to abbreviating it further to:
```
-- the lift is obtained by p⁻¹ ∘ (B → Y)
has_lifting_property i (arrow.mk p.hom) := λ sq,
⟨⟨{ lift := sq.right ≫ p.inv, }⟩⟩
```

To achieve this, we need two improvements:
1. back in `category_theory.arrow`, change the definition of `lift_struct` to
```
/-- A lift of a commutative square is a diagonal morphism making the two triangles commute. -/
@[ext] structure lift_struct {f g : arrow T} (sq : f ⟶ g) :=
(lift : f.right ⟶ g.left)
(fac_left' : f.hom ≫ lift = sq.left . obviously)
(fac_right' : lift ≫ g.hom = sq.right . obviously)

restate_axiom lift_struct.fac_left'
restate_axiom lift_struct.fac_right'
```

2. Extract the goal
```
XY: C
i: arrow C
p: X ≅ Y
sq: i ⟶ arrow.mk p.hom
⊢ i.hom ≫ sq.right ≫ p.inv = sq.left
```
as a `@[simp]` lemma back in `category_theory.arrow`. (And some similar variations.)

These sort of optimisations tend to pay off in the long run, as everyone downstream of you reaps the advantages of a more capable `simp`.